File: day03/day3.hs
module Main where
import Data.String( fromString )
import Data.Text( splitOn, Text, unpack )
import Data.List( intersect )
import Data.Char( ord )


concat3 :: [a] -> [[a]]
concat3 [] = [[]]
concat3 xs = take 3 xs : concat3 (drop 3 xs)

priority :: Char -> Int
priority c
    | ord c <= 90 = ord c - 38
    | ord c > 90 = ord c - 96
    | otherwise = 0


main :: IO ()
main = do
    raw_data <- readFile "input.txt"
    let lines = init $ map unpack (splitOn (fromString "\n") (fromString raw_data))
    let badges = map priority $ init $ map (head . foldl intersect (['a'..'z'] ++ ['A'..'Z'])) (concat3 lines)
    print $ sum badges

File: day03/day3.scala
def priority(c: Char): Int = 
  val z = c.toInt
  return if (z <= 90) z - 38 else z - 96

def commonChar(xs: Seq[String]): Char =
  (('a' to 'z') ++ ('A' to 'Z')).find(c => xs.forall(_.contains(c))).get

@main def hello: Unit = 
  val source = scala.io.Source.fromFile("input.txt")
  val lines = try source.mkString.split("\n") finally source.close()
  println(f"Part 1: ${lines.map(x => priority(commonChar(List(x.take(x.length/2), x.drop(x.length/2))))).sum}")
  println(f"Part 2: ${lines.grouped(3).map(x => priority(commonChar(x))).sum}")
